content/blog/2022-03-21-conservative-gc/index.md

```diff
@@ -0,0 +1,256 @@
+
+<center>
+<img src="fig2.png" alt="Immix Heap Organisation" width="100%" />
+</center>
+
+**Immix** divides the heap into blocks (32 KB) and each block into lines (256 bytes). Objects are
+allocated using bump-pointer allocation within a line. The tracing version of Immix performs
+opportunistic copying to defragment lines and blocks. **Sticky Immix** is a generational
+collector which marks objects that survive collection as old and only performs tracing on young objects.
+**RC Immix** performs deferred reference counting with the Immix heap.
+
+## Making Collectors Conservative
+
+**Object Map** is a data structure proposed by the authors for efficient filtering of ambiguous
+roots. The idea is to maintain a bitmap for each 8-byte word in memory indicating whether that
+location is alive.
+
+**Conservative Immix (Immix<sub>Cons</sub>)** is a conservative version of Immix. It only
+considers roots as ambiguous, and the objects referenced by these are pinned on line
+granularity (256 bytes) which is space efficient. Like Immix, it performs opportunistic
+copying for other non-root exact references. Furthermore, ambiguous roots are filtered
+against the object map, which is updated after every collection cycle.
+
+Similarly, **Conservative Sticky Immix (Sticky Immix<sub>Cons</sub>)** is the conservative
+counterpart of Sticky Immix. Maintaining the object map is trickier since only young objects
+are traced regularly.
+
+**Conservative Immix RC (RC Immix<sub>Cons</sub>)** performs tracing on young objects, and
+reference counting on mature objects. Object map is used here as well, and is updated incrementally
+as reference counting is performed.
+
+# Evaluation
+
+The authors conduct an extensive evaluation in two parts: (1)
+understanding the impact of conservatism by comparing between exact
+and conservative versions of collectors; (2) a performance evaluation
+comparing the conservative versions of RC, Immix, Sticky Immix, and RC
+Immix against existing state-of-the-art conservative garbage
+collectors.
+
+## Impact of Conservatism
+
+The authors evaluated the impacts of conservative-ness on collector
+mechanisms and design, specifically on the number of roots tracked,
+filtering, excess retention, and pinning.
+
+The takeaways are as follows:
+
+- *Ambiguous Pointers*: A challenge in conservative collection is how
+   often a non-pointer object may be interpreted as a pointer. The
+   authors found that conservative scanning results in _60%_
+   more "roots" than exact scanning.
+
+- *Excess Retention*: While excess retention is a very obvious side
+   effect of conservative collection, its precise impacts in practice
+   were not known. The authors measured excess retention by comparing
+   the sizes of transitive closures between the exact and conservative
+   versions, and found that excess retention was, on average, only _0.02%_,
+   with the maximum being only _6.1%_, as a fraction of heap size.
```

Comment:
If you want, this could be a good place to link to [Boehm's theoretical work explaining why excess retention can be surprisingly small](https://www.hpl.hp.com/techreports/2001/HPL-2001-251.pdf).

Reply:
Cited the article.